docs/services.md

```diff
@@ -0,0 +1,84 @@
+# Services
+
+## Abstract
+
+In sct we have repeated pattern when we run code in parallel and time to time need to sync with it, get status, make it stop.
+This pattern is to be called service.
+List key features that distinguish service pattern:
+- Code is run in parallel
+- Code is doing repetitive task
+- You do one of the following:
+    - Check the code status
+    - Make it stop by command
+    - Want it to stop automatically when:
+        - Test is finished
+        - Node is destroyed
+        - Cluster is destroyed
+
+## Goals
+### Effortless lifecycle
+Make services stopped/killed automatically, with no additional code
+### Effortless dependency tracking
+Make services to automatically know their dependencies, so that they could be stopped/killed in proper manner, to avoid unexpected errors to pop up
+### Logging
+Make logging structured so that you could see which service generated an error and which node/cluster associated with it
+### Provide same interface for all threads/processes across sct
+Big problem now when we have tens of threads/processes managed in different manner. It makes hard to do things like:
+- Track them
+- Stop or kill them, especially to do it gracefully
+
+## Type of services
+
+* Types by attachment
+    * Detached Service
+       > Service that is not attached to node or cluster
+Life cycle of this type of service ends when test is finished
+It will be stopped/killed only after all it's dependants are stopped/killed
+
+    * Node Service
+        > Service that is attached to a node.
+Life cycle of this type of service ends when node is destroyed or test is finished
+It will be stopped/killed only after all it's dependants are stopped/killed
+
+    * Cluster Service
+        > Service that is attached to a cluster.
+Life cycle of this type of service ends when cluster is destroyed or test is finished
+It will be stopped/killed only after all it's dependants are stopped/killed
+
+* Types by container
+    * Thread
+        > Service is ran within threading.Thread
+    * Process
+        > Service is ran within multiprocessing.Process
+
+## Interface
+
+Despite types any service has following methods/interfaces:
+
+* start()
+* stop()
+* kill()
+* is_service_alive()
+* cleanup()
+* is_alive()
+* wait_till_stopped(timeout)
+* dependants
+* tags
+* service_name
+
+For more details take a look at sdcm/services/base.py:BaseService
+
+## List of services
+
+This is auxiliary class that provides interface for bulk actions on list of services in chain manner.
```

Comment:
the tagging thing is something that might be of great value.
I have a use case from today:

I have 3 processes in my test for alternator streams:
* `Stress command - YCSB` - run on a loader with docker
* `table clone (a.k.a hydra-kcl)` - a java program that uses kinesis-dynamodb-adapter - run in docker
* `table size compare` - a python process at the test end, that check the number of partitions in a table with `nodetool cfstats`, and when they match, would want to stop `table clone`

Today I don't have any reference to the stress threads in a global place, to kill/stop one of them. it's hidden inside the test themselves.

Also for writing `table size compare`, I've went to the extent of doing something similar to the fullscan, but make it as a stress tool from configuration POV.

see:
https://github.com/scylladb/scylla-cluster-tests/pull/2819/commits/f4074d6187e73337a058131f3f81ea5e48ef3c91
https://github.com/scylladb/scylla-cluster-tests/pull/2819/commits/064755fc6149e39882e61c3ac6a6cc2be9821278

and https://github.com/scylladb/scylla-cluster-tests/pull/2819 in general for and idea of what I'm talking about.